doc/contributing/erm-guidelines.md

```diff
@@ -0,0 +1,453 @@
+   cannot be made disposable without changing the return type or making it
+   polymorphic, consider introducing a new API variant. For example,
+   `fs.mkdtempDisposable` could be introduced to return a disposable object
+   while the existing `fs.mkdtemp` API continues to return a string. Yes, it
+   means more APIs to maintain but it avoids the complexity and confusion of
+   polymorphic return values. If adding a new API variant is not ideal, remember
+   that changing the return type of an existing API is quite likely a breaking
+   change.
+3. When an existing API signature does not lend itself easily to supporting making
+   the return value disposable and a new API needs to be introduced, it is worth
+   considering whether the existing API should be deprecated in favor of the new.
+   Deprecation is never a decision to be taken lightly, however, as it can have major
+   ecosystem impact.
+
+## Guidelines for using disposable objects
+
+Because disposable objects can be disposed of at any time, it is important
+to be careful when using them. Here are some guidelines for using disposable:
+
+1. Never use `using` or `await using` with disposable objects that you
+   do not own. For instance, the following code is problematic if you
+   are not the owner of `someObject`:
+
+```js
+function foo(someObject) {
+  using resource = someObject;
+}
+```
+
+The reason this is problematic is that the `using` statement will
+unconditionally call the `Symbol.dispose` method on `someObject` when the block
+exits, but you do not control the lifecycle of `someObject`. If `someObject`
+is disposed of, it may lead to unexpected behavior in the rest of the
+code that called the `foo` function.
+
+2. When there is a clear difference between disposing of an object in a success
+   context vs. an exception context, always explicitly dispose of objects the
+   successful code paths, including early returns. For example:
+
+```js
+class MyDisposableResource {
+  close() {
+    console.log('Resource closed');
+  }
+
+  abort() {
+    console.log('Resource aborted');
+  }
+
+  [Symbol.dispose]() {
+    // Assume the error case here...
+    this.abort();
+  }
+}
+
+function foo() {
+  using res = new MyDisposableResource();
+  if (someCondition) {
+    // Early return, ensure the resource is disposed of
+    res.close();
+    return;
+  }
+  // do other stuff
+  res.close();
+}
+```
+
+This is because of the fact that, when the disposer is called, it has no way
+of knowing if there is a pending exception or not and it is generally safest
+to assume that it is being called in an exceptional state.
+
+Many types of disposable objects make no differentiation between success and
+exception cases, in which case relying entirely on `using` is just fine (and
+preferred). The disposable returned by `setTimeout()` is a good example here.
+All that does is call `clearTimeout()` and it does not matter if the block
```

Comment:
Ditto above.

Reply:
Switched them to `setImmediate`/`clearImmediate`